Programming assignment #1

TatsianaGouteeva opened the pull request:
Hi!

Please, check my hometask "Programming assignment #1"

BongDiDong commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,54 @@
-class Stack {
-    private var array = [Int]()
+class Stack<T> {
```
I think stack also wanna documentation).

TatsianaGouteeva replied:
done

BongDiDong commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,54 @@
     
-    func push(element: Int) {
+    /**
+     Adds a new element  on the top of the stack.
```
```suggestion
     Adds a new element on the top of the stack.
```

TatsianaGouteeva replied:
done

BongDiDong commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,54 @@
-    private var minElement: Int?
-
-    override func push(element: Int) {
+class StackStatistics<Element: Numeric & Comparable>: Stack<Element> {
```
Please, add doc for this class.

TatsianaGouteeva replied:
done

AndreyKovalevskiy commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -36,7 +63,7 @@ class StackStatistics: Stack {
     }
     
-    override func pop() -> Int? {
+    override func pop() -> Element? {
```
```suggestion
    override func pop() -> Element? {
```
I think it is necessary to add a description for this method

TatsianaGouteeva replied:
done

AndreyKovalevskiy commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,54 @@
+     */
+    private var minElement: Element?
+    
+    override func push(element: Element) {
```
```suggestion
    override func push(element: Element) {
```
I think it is necessary to add a description for this method

TatsianaGouteeva replied:
done

AndreyKovalevskiy commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -52,7 +79,7 @@ class StackStatistics: Stack {
     }
     
-    override func top() -> Int? {
+    override func top() -> Element? {
```
```suggestion
    override func top() -> Element? {
```
And here please add a description for this method

TatsianaGouteeva replied:
done

Hannabolotova commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,63 @@
-    private var array = [Int]()
+/**
+ Class describing a collection of elements in LIFO order.
+ */
```
```suggestion
/*
 Class describing a collection of elements in LIFO order.
 */
```
please, look at documentation about multiline comments. https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html

TatsianaGouteeva replied:
@Hannabolotova I used structure 
`/** 
       ...  
*/` accordint to [Apple's documentation](https://developer.apple.com/library/archive/documentation/Xcode/Reference/xcode_markup_formatting_ref/ComentBlock.html#//apple_ref/doc/uid/TP40016497-CH103-SW1)

dersim-davaod commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,63 @@
+    private var minElement: Element?
+    
+    /**
+     Adds a new element on the top of the stack.
```
nit:

```suggestion
     Adds a new element to the stack.
```

TatsianaGouteeva replied:
done

dersim-davaod commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -36,7 +72,13 @@ class StackStatistics: Stack {
+    /**
+     Removes and returns the top element of the stack.
+     
+     - Returns: The top element of the stack if the stack is not empty; otherwise, nil.
```
nit:

```suggestion
     - Returns: The element of the stack which was added last; If the stack is empty, returns nil.
```

TatsianaGouteeva replied:
done

dersim-davaod commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -62,7 +110,13 @@ class StackStatistics: Stack {
+     Returns the minimum element in the stack.
+     
+     - Returns: The minimum element in the stack if the stack is not empty; otherwise, nil.
+     - Complexity: O(1)
```
nit:

Add info about space-complexity.

TatsianaGouteeva replied:
done

TatsianaGouteeva commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,63 @@
+     - Complexity: O(1)
+     */
+    func pop() -> T? {
         guard !array.isEmpty else {
```
done

TatsianaGouteeva commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,63 @@
+/**
+ Class describing a collection of elements in LIFO order.
+ */
+class StackStatistics<Element: Numeric & Comparable>: Stack<Element> {
```
As I used binary operators (lines #59, #84) so it's possible with types, which support protocol Numeric.

dersim-davaod replied:
So I can't use a `StackStatistics` class for strings?

TatsianaGouteeva replied:
Yes, you can't

dersim-davaod replied:
But I want! :)

TatsianaGouteeva replied:
I deleted condition to work with Numerics. So you can use a StackStatistics class for strings

TatsianaGouteeva commented:
> When we have to document the complexity of properties?

@dersim-davaod  according to [API Design Guidelines](https://swift.org/documentation/api-design-guidelines/)  we have to document the complexity of any computed property that is not O(1).

dersim-davaod commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,59 @@
+ */
+class Stack<T> {
+    /**
+     An array of 'T' elements
```
Is that useful doc?

TatsianaGouteeva replied:
No. But it's not a mistake. I'll delete this just to make code more clear.

dersim-davaod replied:
In that case please give a more clear name to array. It's hard to understand what is its purpose.

TatsianaGouteeva replied:
ok

dersim-davaod commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,59 @@
+     - Parameter element: The element to add to the stack.
+     - Complexity: O(1)
+     */
+    override func push(element: Element) {
```
Seems `push()`/`pop()` methods have the same doc comments as in the basic class?

TatsianaGouteeva replied:
Done. I should follow DRY principle.

dersim-davaod commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,27 +1,59 @@
-        
-        return array.removeLast()
+    /**
+     Removes and returns the top element of the stack.
```
Why you do not use the same description as above?
```
Removes and returns the element, which was added last to the stack
```

I really like it.

TatsianaGouteeva replied:
Done

TatsianaGouteeva commented:
Hi!
Please, check my changes.

- I deleted my condition to work with Numerics. 
- I deleted condition about space complexity O(1) (It wasn't a mandory task)

AndreyKovalevskiy commented on Cocoaheads School_Lesson2.playground/Contents.swift:
```diff
@@ -1,68 +1,64 @@
-class StackStatistics: Stack {
-    private var minElement: Int?
+/**
+ Class describing a collection of elements in LIFO order. Elements of collection must conform to the Numeric and Comparable protocols.
```
```suggestion
 Class describing a collection of elements in LIFO order. Elements of collection must conform to the Comparable protocol.
```

AndreyKovalevskiy replied:
I see that you do not use Numeric protocol

TatsianaGouteeva replied:
Yes, you are right. Thank you!
